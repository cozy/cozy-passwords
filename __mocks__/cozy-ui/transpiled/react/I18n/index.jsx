import React from 'react'

const translate = () => Wrapped => props => {
  return <Wrapped t={key => key} {...props} />
}

export { translate }
