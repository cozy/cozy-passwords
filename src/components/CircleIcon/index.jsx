import React from 'react'
import Icon from 'cozy-ui/transpiled/react/Icon'
import './styles.css'
import cx from 'classnames'

const CircleIcon = props => {
  const { className, ...rest } = props

  return (
    <div className={cx('CircleIcon', className)}>
      <Icon {...rest} />
    </div>
  )
}

export default CircleIcon
