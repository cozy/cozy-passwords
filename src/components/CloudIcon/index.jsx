import React from 'react'
import Icon from 'cozy-ui/transpiled/react/Icon'
import './styles.css'

const CloudIcon = props => {
  return (
    <Icon
      icon="cloud"
      size={16}
      color="var(--charcoalGrey)"
      className="CloudIcon"
      {...props}
    />
  )
}

export default CloudIcon
