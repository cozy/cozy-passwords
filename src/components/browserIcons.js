import BrowserSafariIcon from 'cozy-ui/transpiled/react/Icons/BrowserSafari'
import BrowserFirefoxIcon from 'cozy-ui/transpiled/react/Icons/BrowserFirefox'
import BrowserChromeIcon from 'cozy-ui/transpiled/react/Icons/BrowserChrome'
import BrowserEdgeChromiumIcon from 'cozy-ui/transpiled/react/Icons/BrowserEdgeChromium'

const browserIcons = {
  safari: BrowserSafariIcon,
  firefox: BrowserFirefoxIcon,
  chrome: BrowserChromeIcon,
  'edge-chromium': BrowserEdgeChromiumIcon
}

export default browserIcons
