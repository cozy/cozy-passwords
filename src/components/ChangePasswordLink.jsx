import React from 'react'

import generateWebAppLink from 'helpers/generateWebAppLink'
import AppLinker from 'cozy-ui/transpiled/react/AppLinker'
import { ButtonLink } from 'cozy-ui/transpiled/react/Button'

import { useClient } from 'cozy-client'

const SETTINGS_APP_SLUG = 'settings'
const PASSWORD_APP_SLUG = 'passwords'

const ChangePasswordLink = ({
  successRoute,
  cancelRoute,
  Component,
  ...props
}) => {
  const client = useClient()
  const rawSettingsAppHref = generateWebAppLink(SETTINGS_APP_SLUG, client)
  const passwordsUrl = generateWebAppLink(PASSWORD_APP_SLUG, client)

  const successUrl = new URL(successRoute, passwordsUrl).href
  const cancelUrl = new URL(cancelRoute, passwordsUrl).href

  const settingsPath = '#/profile/password'
  const settingsQuery = `?redirect_success=${encodeURIComponent(
    successUrl
  )}&redirect_cancel=${encodeURIComponent(cancelUrl)}`

  const settingsAppHref = new URL(
    settingsPath + settingsQuery,
    rawSettingsAppHref
  ).href

  return (
    <AppLinker slug={SETTINGS_APP_SLUG} href={settingsAppHref}>
      {({ onClick, href }) => (
        <Component href={href} onClick={onClick} {...props} />
      )}
    </AppLinker>
  )
}

ChangePasswordLink.defaultProps = {
  Component: ButtonLink
}

export default ChangePasswordLink
