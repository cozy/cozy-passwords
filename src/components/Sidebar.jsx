import React, { useContext } from 'react'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import { NavLink } from 'react-router-dom'
import { BitwardenSettingsContext } from '../bitwarden-settings'
import flag from 'cozy-flags'

const isInstallationLinkMatching = (match, location) => {
  if (!match) {
    return false
  }
  if (location.pathname.includes('import')) {
    return false
  }
  return true
}

export const Sidebar = () => {
  const { t } = useI18n()
  const bitwardenSettings = useContext(BitwardenSettingsContext)
  const isVaultConfigured =
    bitwardenSettings && bitwardenSettings.extension_installed
  return isVaultConfigured ? (
    <aside className="o-sidebar">
      <nav>
        <ul className="c-nav">
          <li className="c-nav-item">
            <NavLink
              to="/installation"
              className="c-nav-link"
              isActive={isInstallationLinkMatching}
              activeClassName="is-active"
            >
              {t('Nav.installation')}
            </NavLink>
          </li>
          <li className="c-nav-item">
            <NavLink
              to="/installation/import"
              className="c-nav-link"
              activeClassName="is-active"
            >
              {t('Nav.import')}
            </NavLink>
          </li>
        </ul>
      </nav>
    </aside>
  ) : null
}

export default flag.connect(Sidebar)
