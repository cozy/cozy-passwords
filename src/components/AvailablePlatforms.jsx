import React, { useState, useCallback } from 'react'
import keyBy from 'lodash/keyBy'

import { ButtonLink } from 'cozy-ui/transpiled/react/Button'
import getSupportedPlatforms, { platforms } from 'supportedPlatforms'
import Stack from 'cozy-ui/transpiled/react/Stack'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import Card from 'cozy-ui/transpiled/react/Card'
import Icon from 'cozy-ui/transpiled/react/Icon'
import { Text } from 'cozy-ui/transpiled/react/Text'
import { Dialog } from 'cozy-ui/transpiled/react/CozyDialogs'
import PhoneIcon from 'cozy-ui/transpiled/react/Icons/Phone'

import { isAndroid, isIOS } from 'cozy-device-helper'

import { AppStoreButton, PlayStoreButton } from './StoreButtons'
import browserIcons from './browserIcons'

const PlatformButton = props => {
  const { icon, ...rest } = props
  const color = props.theme === 'primary' ? 'var(--white)' : 'var(--slateGrey)'
  return (
    <ButtonLink
      icon={<Icon icon={icon} size={16} color={color} />}
      theme="secondary"
      className="u-mb-half"
      {...rest}
    />
  )
}
const storeLinksPerOS = keyBy(
  Object.values(platforms).filter(x => x.os),
  'os'
)

export const InstallNativeAppButton = props => {
  const [isSmartphoneModalOpened, setSmartphoneModalOpened] = useState(false)
  const mobileOS = isAndroid() ? 'android' : isIOS() ? 'ios' : null
  const handleOpenModal = useCallback(() => {
    setSmartphoneModalOpened(true)
  }, [setSmartphoneModalOpened])

  const handleDismissModal = useCallback(() => {
    setSmartphoneModalOpened(false)
  }, [setSmartphoneModalOpened])

  return (
    <>
      <PlatformButton
        icon={PhoneIcon}
        href={mobileOS !== null ? storeLinksPerOS[mobileOS].storeUrl : null}
        onClick={mobileOS === null ? handleOpenModal : null}
        {...props}
      />
      {isSmartphoneModalOpened ? (
        <Dialog
          open={isSmartphoneModalOpened}
          onClose={handleDismissModal}
          content={
            <div className="u-flex u-flex-column u-flex-justify-center">
              <div className="u-ta-center">
                <AppStoreButton href={storeLinksPerOS.ios.storeUrl} />
                <PlayStoreButton href={storeLinksPerOS.android.storeUrl} />
              </div>
            </div>
          }
        />
      ) : null}
    </>
  )
}

const AvailablePlatforms = props => {
  const { t } = useI18n()
  const supportedPlatforms = getSupportedPlatforms()
  return (
    <Card {...props}>
      <Stack spacing="m">
        <Text>{t('AvailablePlatforms.text')}</Text>
        <div>
          {Object.entries(supportedPlatforms)
            .filter(([, infos]) => infos.type === 'browser')
            .map(([platform, infos]) => (
              <PlatformButton
                key={platform}
                href={infos.storeUrl}
                icon={browserIcons[platform]}
                label={infos.label}
              />
            ))}
          <InstallNativeAppButton label={t('AvailablePlatforms.smartphone')} />
        </div>
      </Stack>
    </Card>
  )
}

export default AvailablePlatforms
