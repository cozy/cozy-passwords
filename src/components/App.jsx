import React, { useEffect } from 'react'
import { hot } from 'react-hot-loader'
import { Route, Switch, Redirect, HashRouter } from 'react-router-dom'
import { Layout, Main, Content } from 'cozy-ui/transpiled/react/Layout'
import { Sprite as IconSprite } from 'cozy-ui/transpiled/react/Icon'
import Alerter from 'cozy-ui/transpiled/react/Alerter'
import MuiCozyTheme from 'cozy-ui/transpiled/react/MuiCozyTheme'
import withBreakpoints from 'cozy-ui/transpiled/react/helpers/withBreakpoints'

import Sidebar from './Sidebar'
import PresentationPage from './PresentationPage'
import SecurityPage from './SecurityPage'
import HintPage from './HintPage'
import InstallationPage from './InstallationPage'
import InstalledPage from './InstalledPage'
import NotAvailablePage from './NotAvailablePage'
import ConnectedPage from './ConnectedPage'
import ImportPage from './ImportPage'
import {
  useExtensionStatus,
  extensionStatuses
} from '../helpers/extensionStatus'
import flag, { FlagSwitcher } from 'cozy-flags'
import minilog from 'minilog'

window.minilog = minilog
window.flag = flag

const RedirectIfExtensionInstalledOrConnected = props => {
  const extensionStatus = useExtensionStatus()

  if (extensionStatus === extensionStatuses.checking) {
    return null
  }

  if (extensionStatus === extensionStatuses.installed) {
    return <Redirect to="/installation/installed" />
  }

  if (extensionStatus === extensionStatuses.connected) {
    return <Redirect to="/installation/connected" />
  }

  return <Route {...props} />
}

/*
 * This is not very DRY to have RedirectIfExtensionInstalledOrConnected and
 * RedirectIfExtensionConnected, but the problem is that on
 * /installation/installed route we can't listen for the installed status or we
 * will have an infinite redirection loop. So in this case we just want to
 * listen for connection.
 */
const RedirectIfExtensionConnected = props => {
  const extensionStatus = useExtensionStatus()

  if (extensionStatus === extensionStatuses.checking) {
    return null
  }

  if (extensionStatus === extensionStatuses.connected) {
    return <Redirect to="/installation/connected" />
  }

  return <Route {...props} />
}

export const DumbApp = ({ breakpoints: { isDesktop } }) => {
  useEffect(() => {
    if (process.env.NODE_ENV === 'development') {
      flag('switcher', true)
    }
  }, [])

  if (isDesktop) {
    return (
      <MuiCozyTheme>
        <HashRouter>
          <Layout>
            <Sidebar />
            <Main>
              <Content>
                <Switch>
                  <RedirectIfExtensionInstalledOrConnected
                    path="/presentation"
                    component={PresentationPage}
                  />
                  <RedirectIfExtensionInstalledOrConnected
                    path="/security"
                    exact
                    component={SecurityPage}
                  />
                  <RedirectIfExtensionInstalledOrConnected
                    path="/security/hint"
                    exact
                    component={HintPage}
                  />
                  <RedirectIfExtensionInstalledOrConnected
                    path="/installation"
                    exact
                    component={InstallationPage}
                  />
                  <RedirectIfExtensionConnected
                    path="/installation/installed"
                    exact
                    component={InstalledPage}
                  />
                  <Route
                    path="/installation/connected"
                    exact
                    component={ConnectedPage}
                  />
                  <Route
                    path="/installation/import"
                    exact
                    component={ImportPage}
                  />
                  <Redirect from="/" to="/presentation" />
                  <Redirect from="*" to="/presentation" />
                </Switch>
              </Content>
            </Main>
            <IconSprite />
            <Alerter />
            <FlagSwitcher />
          </Layout>
        </HashRouter>
      </MuiCozyTheme>
    )
  }

  return (
    <>
      <NotAvailablePage />
      <IconSprite />
    </>
  )
}

const App = withBreakpoints()(DumbApp)

/*
  Enable Hot Module Reload using `react-hot-loader` here
  We enable it here since App is the main root component
  No need to use it anywhere else, it sould work for all
  child components
*/
export default hot(module)(App)
