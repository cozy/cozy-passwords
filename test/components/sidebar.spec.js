'use strict'

/* eslint-env jest */

import React from 'react'
import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'

import AppLike from '../lib/AppLike'
import { Sidebar } from 'components/Sidebar'

describe('Sidebar component', () => {
  it('should be rendered correctly', () => {
    const { getByText } = render(
      <AppLike>
        <MemoryRouter>
          <Sidebar />
        </MemoryRouter>
      </AppLike>
    )

    expect(getByText(/presentation/i)).toBeDefined()
    expect(getByText(/security/i)).toBeDefined()
    expect(getByText(/installation/i)).toBeDefined()
  })
})
