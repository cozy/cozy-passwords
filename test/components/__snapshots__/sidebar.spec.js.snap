// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component should be rendered correctly 1`] = `
<aside
  className="o-sidebar"
>
  <nav>
    <ul
      className="c-nav"
    >
      <li
        className="c-nav-item"
      >
        <ForwardRef
          activeClassName="is-active"
          className="c-nav-link"
          to="/presentation"
        >
          <span
            className="c-nav-number"
          >
            1.
          </span>
          Presentation
        </ForwardRef>
      </li>
      <li
        className="c-nav-item"
      >
        <ForwardRef
          activeClassName="is-active"
          className="c-nav-link"
          to="/security"
        >
          <span
            className="c-nav-number"
          >
            2.
          </span>
          Security
        </ForwardRef>
      </li>
      <li
        className="c-nav-item"
      >
        <ForwardRef
          activeClassName="is-active"
          className="c-nav-link"
          to="/installation"
        >
          <span
            className="c-nav-number"
          >
            3.
          </span>
          Installation
        </ForwardRef>
      </li>
    </ul>
  </nav>
</aside>
`;
